File: android/src/main/java/app/capgo/recaptcha/RecaptchaPlugin.java
package app.capgo.recaptcha;

import android.app.Application;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "Recaptcha")
public class RecaptchaPlugin extends Plugin {

    private final Recaptcha implementation = new Recaptcha();

    @PluginMethod
    public void load(PluginCall call) {
        String siteKey = resolveSiteKey(call);
        if (isBlank(siteKey)) {
            call.reject("siteKey is required. Pass siteKey/androidSiteKey or set Recaptcha.androidSiteKey/siteKey in Capacitor config.");
            return;
        }

        implementation.load(
            getApplication(),
            siteKey,
            new Recaptcha.LoadCallback() {
                @Override
                public void onSuccess() {
                    call.resolve(loadResult(siteKey));
                }

                @Override
                public void onFailure(Exception error) {
                    call.reject("Failed to load reCAPTCHA.", "LOAD_FAILED", error);
                }
            }
        );
    }

    @PluginMethod
    public void execute(PluginCall call) {
        String siteKey = resolveSiteKey(call);
        if (isBlank(siteKey)) {
            call.reject("siteKey is required. Pass siteKey/androidSiteKey or set Recaptcha.androidSiteKey/siteKey in Capacitor config.");
            return;
        }

        String action = call.getString("action");
        if (isBlank(action)) {
            call.reject("action is required.");
            return;
        }

        Long timeout = call.getLong("timeout");

        implementation.execute(
            getApplication(),
            siteKey,
            action,
            timeout,
            new Recaptcha.ExecuteCallback() {
                @Override
                public void onSuccess(String token) {
                    JSObject ret = executionResult(token, action.trim(), siteKey);
                    call.resolve(ret);
                }

                @Override
                public void onFailure(Exception error) {
                    call.reject(Recaptcha.errorMessage(error), "EXECUTE_FAILED", error);
                }
            }
        );
    }

    @PluginMethod
    public void getPluginVersion(PluginCall call) {
        JSObject ret = new JSObject();
        ret.put("version", implementation.getPluginVersion());
        call.resolve(ret);
    }

    private Application getApplication() {
        return (Application) getContext().getApplicationContext();
    }

    private JSObject loadResult(String siteKey) {
        JSObject ret = new JSObject();
        ret.put("loaded", true);
        ret.put("siteKey", siteKey);
        ret.put("enterprise", true);
        ret.put("platform", "android");
        return ret;
    }

    private JSObject executionResult(String token, String action, String siteKey) {
        JSObject ret = new JSObject();
        ret.put("token", token);
        ret.put("action", action);
        ret.put("siteKey", siteKey);
        ret.put("enterprise", true);
        ret.put("platform", "android");
        return ret;
    }

    private String resolveSiteKey(PluginCall call) {
        return firstNonBlank(
            call.getString("androidSiteKey"),
            call.getString("siteKey"),
            call.getString("sitekeyAndroid"),
            getConfig().getString("androidSiteKey"),
            getConfig().getString("siteKey")
        );
    }

    private String firstNonBlank(String... values) {
        for (String value : values) {
            if (!isBlank(value)) {
                return value.trim();
            }
        }
        return null;
    }

    private boolean isBlank(String value) {
        return value == null || value.trim().isEmpty();
    }
}

File: ios/Sources/RecaptchaPlugin/RecaptchaPlugin.swift
import Foundation
import Capacitor
import RecaptchaEnterprise

@objc(RecaptchaPlugin)
public class RecaptchaPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "RecaptchaPlugin"
    public let jsName = "Recaptcha"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "load", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "execute", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getPluginVersion", returnType: CAPPluginReturnPromise)
    ]

    private let implementation = RecaptchaBridge()

    @objc func load(_ call: CAPPluginCall) {
        guard resolveEnterprise(call) else {
            rejectStandardNativeMode(call)
            return
        }

        guard let siteKey = resolveSiteKey(call) else {
            call.reject("siteKey is required. Pass siteKey/iosSiteKey or set Recaptcha.iosSiteKey/siteKey in Capacitor config.")
            return
        }

        Task {
            do {
                try await implementation.load(siteKey: siteKey)
                call.resolve(loadResult(siteKey: siteKey))
            } catch {
                reject(call, error: error, fallbackMessage: "Failed to load reCAPTCHA.")
            }
        }
    }

    @objc func execute(_ call: CAPPluginCall) {
        guard resolveEnterprise(call) else {
            rejectStandardNativeMode(call)
            return
        }

        guard let siteKey = resolveSiteKey(call) else {
            call.reject("siteKey is required. Pass siteKey/iosSiteKey or set Recaptcha.iosSiteKey/siteKey in Capacitor config.")
            return
        }

        guard let action = call.getString("action")?.trimmingCharacters(in: .whitespacesAndNewlines), !action.isEmpty else {
            call.reject("action is required.")
            return
        }

        let timeout = call.getInt("timeout")

        Task {
            do {
                let token = try await implementation.execute(siteKey: siteKey, action: action, timeout: timeout)
                call.resolve(executionResult(token: token, action: action, siteKey: siteKey))
            } catch {
                reject(call, error: error, fallbackMessage: "Failed to execute reCAPTCHA.")
            }
        }
    }

    @objc func getPluginVersion(_ call: CAPPluginCall) {
        call.resolve([
            "version": implementation.getPluginVersion()
        ])
    }

    private func resolveSiteKey(_ call: CAPPluginCall) -> String? {
        firstNonEmpty(
            call.getString("iosSiteKey"),
            call.getString("siteKey"),
            call.getString("sitekeyIos"),
            call.getString("sitekeyIOS"),
            getConfig().getString("iosSiteKey"),
            getConfig().getString("siteKey"),
            getConfig().getString("sitekeyIos"),
            getConfig().getString("sitekeyIOS")
        )
    }

    private func resolveEnterprise(_ call: CAPPluginCall) -> Bool {
        call.getBool("enterprise") ?? getConfig().getBoolean("enterprise", true)
    }

    private func rejectStandardNativeMode(_ call: CAPPluginCall) {
        call.reject(
            "Regular reCAPTCHA v3 is only supported on Web. iOS uses Google's mobile reCAPTCHA SDK, which requires an Enterprise/mobile site key.",
            "UNSUPPORTED_MODE"
        )
    }

    private func loadResult(siteKey: String) -> [String: Any] {
        [
            "loaded": true,
            "siteKey": siteKey,
            "enterprise": true,
            "platform": "ios"
        ]
    }

    private func executionResult(token: String, action: String, siteKey: String) -> [String: Any] {
        [
            "token": token,
            "action": action,
            "siteKey": siteKey,
            "enterprise": true,
            "platform": "ios"
        ]
    }

    private func firstNonEmpty(_ values: String?...) -> String? {
        for value in values {
            guard let trimmedValue = value?.trimmingCharacters(in: .whitespacesAndNewlines), !trimmedValue.isEmpty else {
                continue
            }
            return trimmedValue
        }
        return nil
    }

    private func reject(_ call: CAPPluginCall, error: Error, fallbackMessage: String) {
        if let recaptchaError = error as? RecaptchaError {
            call.reject(
                recaptchaError.errorMessage,
                String(describing: recaptchaError.errorCode),
                recaptchaError
            )
            return
        }

        call.reject(error.localizedDescription.isEmpty ? fallbackMessage : error.localizedDescription, nil, error)
    }
}
